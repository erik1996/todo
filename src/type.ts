export interface IItem {
  id: number;
  title: string;
  status: StatusEnum;
}

export enum StatusEnum {
  TO_DO = "To DO",
  IN_PROGRESS = "In Progress",
  DONE = "Done",
}
