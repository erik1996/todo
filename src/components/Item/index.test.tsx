import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Item } from ".";
import { StatusEnum } from "../../type";
describe("Item", () => {
  const mockItem = {
    id: 1,
    title: "New Item",
    status: StatusEnum.TO_DO,
    handleUpdate: jest.fn(),
    handleRemove: jest.fn(),
  };

  describe("Snapshot testing for TO_DO status", () => {
    test("renders correctly", () => {
      const component = renderer.create(
        <Item
          title={mockItem.title}
          status={StatusEnum.TO_DO}
          id={mockItem.id}
          handleRemove={() => {}}
          handleUpdate={() => {}}
        />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Snapshot testing for IN_PROGRESS status", () => {
    test("renders correctly", () => {
      const component = renderer.create(
        <Item
          title={mockItem.title}
          status={StatusEnum.IN_PROGRESS}
          id={mockItem.id}
          handleRemove={() => {}}
          handleUpdate={() => {}}
        />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  test("should render the item with correct title and background color", () => {
    render(
      <Item
        title={mockItem.title}
        status={mockItem.status}
        id={mockItem.id}
        handleRemove={mockItem.handleRemove}
        handleUpdate={mockItem.handleUpdate}
      />
    );

    const item = screen.getByText(mockItem.title);
    expect(item).toBeInTheDocument();

    const itemContainer = screen.getByRole("listitem");
    expect(itemContainer).toHaveStyle(`background: yellow;`);
  });

  test("should have yellow background color when status is TO_DO", () => {
    render(
      <Item
        title={mockItem.title}
        status={StatusEnum.TO_DO}
        id={mockItem.id}
        handleRemove={() => {}}
        handleUpdate={() => {}}
      />
    );

    const itemContainer = screen.getByRole("listitem");
    expect(itemContainer).toHaveStyle("background: yellow;");
  });

  test("should have yellow background color when status is IN_PROGRESS", () => {
    render(
      <Item
        title={mockItem.title}
        status={StatusEnum.IN_PROGRESS}
        id={mockItem.id}
        handleRemove={() => {}}
        handleUpdate={() => {}}
      />
    );

    const itemContainer = screen.getByRole("listitem");
    expect(itemContainer).toHaveStyle("background: blue;");
  });

  test("should have green background color when status is DONE", () => {
    render(
      <Item
        title={mockItem.title}
        status={StatusEnum.DONE}
        id={mockItem.id}
        handleRemove={() => {}}
        handleUpdate={() => {}}
      />
    );

    const itemContainer = screen.getByRole("listitem");
    expect(itemContainer).toHaveStyle("background: green;");
  });

  test("should call handleUpdate when the status is changed", () => {
    render(
      <Item
        title={mockItem.title}
        status={mockItem.status}
        id={mockItem.id}
        handleRemove={mockItem.handleRemove}
        handleUpdate={mockItem.handleUpdate}
      />
    );

    const selectElement = screen.getByLabelText("status");
    fireEvent.change(selectElement, {
      target: { value: StatusEnum.IN_PROGRESS },
    });

    expect(mockItem.handleUpdate).toHaveBeenCalledWith(
      mockItem.id,
      StatusEnum.IN_PROGRESS
    );
  });

  test("should call handleRemove when click remove button", () => {
    render(
      <Item
        title={mockItem.title}
        status={mockItem.status}
        id={mockItem.id}
        handleRemove={mockItem.handleRemove}
        handleUpdate={mockItem.handleUpdate}
      />
    );

    const selectElement = screen.getByText("Remove");
    fireEvent.click(selectElement);

    expect(mockItem.handleRemove).toHaveBeenCalledWith(mockItem.id);
  });
});
