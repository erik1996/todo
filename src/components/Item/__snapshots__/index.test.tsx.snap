// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Snapshot testing for IN_PROGRESS status renders correctly 1`] = `
<li
  className="item-container"
  style={
    Object {
      "background": "blue",
    }
  }
>
  <span>
    New Item
  </span>
  <select
    aria-label="status"
    onChange={[Function]}
  >
    <option
      value="To DO"
    >
      To DO
    </option>
    <option
      value="In Progress"
    >
      In Progress
    </option>
    <option
      value="Done"
    >
      Done
    </option>
  </select>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</li>
`;

exports[`Item Snapshot testing for TO_DO status renders correctly 1`] = `
<li
  className="item-container"
  style={
    Object {
      "background": "yellow",
    }
  }
>
  <span>
    New Item
  </span>
  <select
    aria-label="status"
    onChange={[Function]}
  >
    <option
      value="To DO"
    >
      To DO
    </option>
    <option
      value="In Progress"
    >
      In Progress
    </option>
    <option
      value="Done"
    >
      Done
    </option>
  </select>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</li>
`;
